src/meta_schedule/integration.cc

```diff
@@ -121,14 +113,18 @@ ApplyHistoryBest::ApplyHistoryBest(Database database) {
   data_ = n;
 }
 
-Optional<ObjectRef> ApplyHistoryBestNode::Query(runtime::String task_name, IRModule mod,
-                                                Target target,
-                                                Optional<Array<IRModule>> dispatched) {
+IRModule ApplyHistoryBestNode::Query(runtime::String task_name, IRModule mod, Target target,
+                                     Optional<Array<IRModule>> dispatched) {
   ICHECK(dispatched.defined());
   ICHECK_EQ(dispatched.value().size(), 1);
   ICHECK(HasOnlyOneFunction<relay::Function>(mod)) << mod;
   IRModule prim_mod = dispatched.value()[0];
   ICHECK(HasOnlyOneFunction<tir::PrimFunc>(prim_mod)) << prim_mod;
+  // TODO(masahi): parse_mod below replaces the orginal function key with "main".
+  // This is necessary because some scheduling primitives requires the PrimFunc key be "main".
```

Comment:
> some scheduling primitives requires the PrimFunc key be "main"

Are you referring to `GetBlock` whose default `func_name` is "main"? We could actually use other names without problem though

Reply:
Yeah interesting, didn't know that. But I don't think it is reasonable to require that every use of `get_block` be given the exact function name rather than `main`. So in practice, we had to use `Parse._mod` thing, which will simplify the use of `get_block` but make this function more complicated than necessary. I updated the TODO comment to be more precise. https://github.com/apache/tvm/pull/10578/commits/5c8800c8d32b27b12cf571e3a4e98767433e0345

To remove `parse_mod` here, which I'd love to do, we also need to remove `Parse._mod` before tuning and require the exact function name to be passed to `get_block`. What do you think?

Reply:
Is this `Lookup` strictly necessary? https://github.com/apache/tvm/blob/96416c4941df0ba292dcf2f7ddb35c8909860341/src/tir/schedule/primitive/get_block_loop.cc#L42

I think we only have one function during tuning.

Reply:
also cc @zxybazh. I'd love to remove the need for `Parse._mod` thing https://github.com/apache/tvm/blob/1271e6cc34eda7be8ce7502c206b1dcef11ab4f1/python/tvm/meta_schedule/tune.py#L214-L219

Is there another reason we need this, other than `get_block` assuming the func name be `main` by default?

Reply:
Thank Masa! To provide more context, MetaSchedule was originally designed to work with only one PrimFunc, but later we decided to potentially generalize it to IRModule-based in case we want to support multi-PrimFunc and mutual function calls in the future.

Reply:
It's indeed weird but true that different global var names lead to failure of structural equality:

https://github.com/apache/tvm/blob/b5f1dabce45e4d4a68a2ecadeee811bac93b1c3c/src/ir/module.cc#L79-L82

, and thus if it isn't taken care of, name mismatches could happen at BaseFunc-level.

On the bright side, if we are able to find an IRModule in the database, where equality is defined as structural equality, this means that we do have the fact that names of the global vars match exactly according to the definition, which means the `func_name` parameter of `GetBlock` should work as it is supposed to be.

Reply:
Hi Masa, thanks for the notice. This is here for another reason that I found during relay tuning, the func name after task extraction are with prefix `vm_mod` and during `ApplyHistoryBest` the func name is `tvm_default`, which might cause a structural unequal when I search for the right candidate in Meta Schedule database. This was a work_around for that.

Reply:
@zxybazh right, now that task extraction is decoupled from VM with this PR, we can revisit this problem (we don't have `vm_mod ` prefix in task name anymore). I think, the key is to make the following two `GlobalVar` identical:

https://github.com/apache/tvm/blob/f30d2bbd643863c3e00b3490c2ef77cbbcbe0da7/src/relay/backend/te_compiler_cache.cc#L310
https://github.com/apache/tvm/blob/f30d2bbd643863c3e00b3490c2ef77cbbcbe0da7/src/relay/backend/task_extraction.cc#L64

Right now, they are not. I thought about this for a bit, making sure they are identical and stay in sync is probably doable but tricky. `Parse._mod` is a bit of hack bit it does simplify the situation.

Reply:
On the dark side, if the global var's name actually differ between task extraction and schedule injection, then we are basically doomed because the system is never able to select anything provided by MetaSchedule

Reply:
yeah, I remember Mark is talking about "unified compiler backend" where there is little distinction between task extraction and compilation. With that we can make sure that there is always unique global var used by task extraction and final compile.

Ironically, the purpose of this PR is to decouple task extraction from VM compile haha... (which is still a huge improvement over the current situation imo)

Reply:
ok I removed the TODO comment above and accept `Parse._mod` as a necessary evil for now.

Reply:
Thanks @masahi and @zxybazh for the discussion! Let's leave `Parse._mod` as it is now and consider refactoring later :-)